### PR TITLE
Darwin support for Predator

### PR DESCRIPTION
**Pull request by versokova**
Add regre tests for llvm: pass only 52%. I don't want to have a conflict with Marek. LLVM IR was changed and builtin function too.

Add regre tests for gcc plug-in:

__builtin_expect() causes an undefined condition. Is the way to eliminate branch prediction?
	14 - test-0014.c (Failed)
	167 - test-0174.c (Failed)

different line:
```
-test-0090.c:9: warning: memory leak detected while destroying a variable on stack
+test-0090.c:21: warning: memory leak detected while destroying a variable on stack
```
	89 - test-0090.c (Failed) 

When Predator reached ERROR label, it catched run-time exception and then gcc-6: internal compiler error: Abort trap: 6 (program cc1). I guess, problem is in brew installation of gcc on mac.
	171 - test-0178.c (Failed)
	173 - test-0180.c (Failed)
	175 - test-0182.c (Failed)
	177 - test-0184.c (Failed)
	179 - test-0186.c (Failed)
	180 - test-0187.c (Failed)
	182 - test-0189.c (Failed)

slgccv and slllvmv doesn't work due to `readlink -f`.

Tested on Darwin 17.3.0 High Sierra with compiler Apple LLVM version 10.0.0 for gcc-6.5.0 and llvm-6.0.1, llvm-7.0.1, llvm-8.0.0.

### CODE REVIEW

File: cl/cl_chain.cc [hunk 1]
```diff
@@ -218,7 +218,7 @@ void cl_chain_append(
 {
     try {
         ICodeListener *listener = cl_obtain_from_wrap(self);
-        ClChain *chain = dynamic_cast<ClChain *>(listener);
+        ClChain *chain = static_cast<ClChain *>(listener);
```
Comment (kdudka) [inline cl/cl_chain.cc]: Does not it indicate some problem with compiler flags?  Perhaps missing RTTI support?

Anyway, the error message for the subsequent  condition `(!chain)` needs to be updated (because `static_cast` never fails at run time).

File: sl/symbin.cc [hunk 2]
```diff
@@ -271,7 +271,8 @@ bool handleNoOp(
     const CodeStorage::TOperandList &opList = insn.operands;
 
     // this allows functions with 0 or 1 parameters
```
Comment (kdudka) [inline sl/symbin.cc]: The comment should be updated to reflect the code.

File: sl/symbin.cc [hunk 3]
```diff
@@ -271,7 +271,8 @@ bool handleNoOp(
...
-    if (opList.size() < 2 || 3 < opList.size()) {
+    if ((opList.size() < 2 || 3 < opList.size())
```
Comment (kdudka) [inline sl/symbin.cc]: Could the condition be simplified as `(opList.size() < 2 || 4 < opList.size())`?

File: README [hunk 4]
```diff
@@ -7,11 +7,11 @@ predicates, but they are now purely graph-based and significantly extended to
 support various forms of low-level memory manipulation used in system-level
 code.  Such operations include pointer arithmetic, safe usage of invalid
 pointers, block operations with memory, reinterpretation of the memory contents,
-address alignment, etc.  The tool can be loaded into *GCC* as a *plug-in*.  This
-way you can easily analyse C code sources, using the existing build system,
-without manually preprocessing them first.  The analysis itself is, however, not
-ready for complex projects yet.  You can find the latest news on the following
-project page:
+address alignment, etc.  The tool can be loaded into *GCC* or *LLVM* as
```
Comment (kdudka) [inline README]: I would postpone changing this part of README until the LLVM support is reasonably stable.

File: README [hunk 5]
```diff
@@ -94,6 +95,58 @@ running the following script:
 
     echo please help | gcc -fplugin=libfa.so -fplugin-arg-libfa-help -xc -
 
+Building from sources against the system LLVM
```
Comment (kdudka) [inline README]: Please make it clear that the LLVM support is experimental and wrap the lines at 80th column.

File: sl/symbin.cc [hunk 6]
```diff
@@ -806,6 +806,34 @@ bool handleStrncpy(
...
+    SymHeap &sh = core.sh();
+    CL_DEBUG_MSG(&insn.loc, "executing " << name << "()");
+
+    // set the returned value to a condition
+    const struct cl_operand &opDst = opList[0];
+    const FldHandle fldDst = core.fldByOperand(opDst);
+    core.setValueOf(fldDst, val);
+
+    // insert the resulting heap
+    dst.insert(sh);
```
Comment (kdudka) [inline sl/symbin.cc]: Please use the `insertCoreHeap()` helper function here.

File: build-aux/common.cmake [hunk 7]
```diff
@@ -15,11 +15,17 @@
 # You should have received a copy of the GNU General Public License
 # along with predator.  If not, see <http://www.gnu.org/licenses/>.
 
-# uncomment this on Darwin if linking the plug-ins fails on undefined references
-#set(CMAKE_SHARED_LINKER_FLAGS "-flat_namespace -undefined suppress")
+# on Darwin linking the plug-ins fails on undefined references
+if(APPLE)
+    set(CMAKE_SHARED_LINKER_FLAGS "-flat_namespace -undefined suppress")
+    # or better used cmake >=3.0
```
[the remark this answers is not in the document]
Comment (versokova) [reply build-aux/common.cmake]: [https://cmake.org/cmake/help/v3.0/prop_tgt/MACOSX_RPATH.html](https://cmake.org/cmake/help/v3.0/prop_tgt/MACOSX_RPATH.html)
On MACOSX it add some more shared libraries into cmakelists, which can be automatically found on linux. In the predator project, are all the necessary llvm libraries explicitly linked. MACOSX_RPATH is set automatically in cmake >=3.0. But I didn't find the equivalent for `get_property(SL_PLUG TARGET sl PROPERTY LOCATION)` in cmake 3.0.

[the remark this answers is not in the document]
Comment (kdudka) [reply build-aux/common.cmake]: Thank you for clarifying it!  So we should also add `cmake_minimum_required(VERSION 3.0)` to this branch to make it fail early with older cmake versions.

[the remark this answers is not in the document]
Comment (versokova) [reply build-aux/common.cmake]: ok, I must change `get_property` too because of errror

File: cl/Makefile.chk [hunk 8]
```diff
@@ -15,15 +15,18 @@
...
 
-.PHONY: check $(HEADERS)
+.PHONY: check $(HEADERS) $(HHEADERS)
 
 # check whether each header is self-compilable
-check: $(HEADERS)
+check: $(HEADERS) $(HHEADERS)
 $(HEADERS):
+	$(CC) $(CFLAGS) -o $(SINK) -c $@
+$(HHEADERS):
```
[the remark this answers is not in the document]
Comment (versokova) [reply cl/Makefile.chk]: I like it. It is more consistent" gcc -> h, g++ -> hh

File: cl/cl_chain.cc (hunk 1 above)
Comment (versokova) [reply cl/cl_chain.cc]:
```
2019-04-20 17:02:27.895012+0200 cc1[41185:895562] dynamic_cast error 1: Both of the following type_info's should have public visibility.  At least one of them is hidden. 
```
RTTI works OK with gcc, in shared library doesn't. `-frtti` is enabled, I try set default visibility, then add option `-fvisibility-inlines-hidden`.
Honestly, I don't know. This applies to all `dynamic_cast`s, but predator works correctly (?!) And each throwing exception ends with abort (no catching unknown exception, all of them) and only, if it's part of shared library.

File: sl/check-property.sh.in [hunk 9]
```diff
@@ -138,7 +143,7 @@ fi; if [ -z "$TASK" ]; then
 fi
 
 # include common code base
-topdir="`dirname "$(readlink -f "$SELF")"`/.."
+topdir="`dirname "$(python -c "import os; print os.path.abspath('$SELF')")"`/.."
```
[the remark this answers is not in the document]
Comment (versokova) [reply sl/check-property.sh.in]: It works, thank you.

File: build-aux/common.cmake [hunk 10]
```diff
@@ -195,7 +203,13 @@ macro(CL_BUILD_COMPILER_PLUGIN PLUGIN ANALYZER LIBCL_PATH)
     endif()
 
     # this will recursively pull all needed symbols from the static libraries
-    set_target_properties(${PLUGIN} PROPERTIES LINK_FLAGS -Wl,--entry=plugin_init)
+    if(NOT APPLE)
+        set_target_properties(${PLUGIN} PROPERTIES LINK_FLAGS -Wl,--entry=plugin_init)
+    else()
+		# on Darwin entry symbol doesn't work properly, load everything
```
[the remark this answers is not in the document]
Comment (versokova) [reply build-aux/common.cmake]: Yes, but of course not for clang. The difference is only 80kB.

File: sl/slscript.in [hunk 11]
```diff
@@ -194,11 +202,12 @@ wait "$!"
 EC="$?"
 test 0 = "$EC" || exit $EC
 
+# FIXME: 'timeout' is not available on Darwin
```
Comment (kdudka) [inline sl/slscript.in]: Looks like some leftover.  Isn't it exactly what this commit fixes?

File: sl/slscript.in [hunk 12]
```diff
@@ -149,8 +158,7 @@ run_tool() {
 
 # colorize the output
 # FIXME: 'stdbuf' is not available on Darwin
```
Comment (kdudka) [inline sl/slscript.in]: This is being fixed, isn't it? Should be safe to remove the stdbuf FIXME, too.

File: cl/Makefile.chk [hunk 13]
```diff
@@ -15,15 +15,18 @@
...
 
-.PHONY: check $(HEADERS)
+.PHONY: check $(C_HEADERS) $(CXX_HEADERS)
 
 # check whether each header is self-compilable
-check: $(HEADERS)
-$(HEADERS):
+check: $(C_HEADERS) $(CXX_HEADERS)
+$(C_HEADERS):
+	$(CC) $(CFLAGS) -o $(SINK) -c $@
```
Comment (kdudka) [inline cl/Makefile.chk]: Could we check `$(C_HEADERS)` by both `$(CC)` and `$(CXX)`?

File: README [hunk 14]
```diff
@@ -38,7 +39,7 @@ GCC.  However, the Predator plug-in itself is known to work with GCC 4.[5-9].x,
 (1) Install all dependences of Predator and Forester
 ----------------------------------------------------
    * C/C++ compiler
-     - Predator currently supports GCC 4.1.2+ and Clang 3.2+
```
Comment (kdudka) [inline README]: Is this being removed on purpose?

I am able to build Predator as a GCC plugin with `clang-3.4.2` and it seems to work properly.

File: cl/CMakeLists.txt [hunk 15]
```diff
@@ -16,7 +16,7 @@
 # along with predator.  If not, see <http://www.gnu.org/licenses/>.
 
 # project metadata
-cmake_minimum_required(VERSION 2.6)
+cmake_minimum_required(VERSION 3.0)
```
Comment (kdudka) [inline cl/CMakeLists.txt]: I think this should be conditioned by `if(APPLE)` in order not to break the build on Ubuntu 14 LTS:
https://travis-ci.org/kdudka/predator/jobs/524427093

File: README (hunk 14 above)
Comment (versokova) [reply README]: Yes, that part is split to two parts 'Building from sources against the system XXX'

File: cl/Makefile.chk (hunk 13 above)
Comment (versokova) [reply cl/Makefile.chk]: But I like it as it is. It is more consistent" gcc -> h, g++ -> hh

File: cl/CMakeLists.txt (hunk 15 above)
Comment (versokova) [reply cl/CMakeLists.txt]: Really? Support ends 2019-04-30.

File: cl/Makefile.chk (hunk 13 above)
Comment (kdudka) [reply cl/Makefile.chk]: The C header files are meant to be valid C and valid C++ at the same time.  This is difficult to check by single invocation of a compiler.

File: cl/CMakeLists.txt (hunk 15 above)
Comment (kdudka) [reply cl/CMakeLists.txt]: According to https://wiki.ubuntu.com/Releases support for Ubuntu 14 LTS goes EOL in April 2022.  Nevertheless I am not against migrating the Travis CI tests to a more recent OS.  According to https://sv-comp.sosy-lab.org/2019/rules.php SV-COMP runs on Ubuntu 18.04 now.

File: README (hunk 14 above)
Comment (kdudka) [reply README]: But this line does not say which compiler Predator can be loaded into.  It says which compiler Predator can be compiled *with*.  One can compile Predator using clang as the compiler and then load the resulting binary into GCC to run the analysis.

Comment (versokova) [reply README]: Yes, I know. But for later version of clang it crash on dry-run regre test (self compilation) of Code Listener. I can't tell from what version it is going on. I have installed last three and it crashed on clang 6 (for gcc-plugin).

Comment (kdudka) [reply README]: I tried it now -- compiled Predator using `clang-6.0.1` against system `gcc-6.4.0` and it worked just fine:
```
$ CC=clang-6.0 CXX=clang++-6.0 ./switch-host-gcc.sh /usr/bin/gcc-6.4.0 
[...]
Trying to build Code Listener...

make: Entering directory '/home/kdudka/git/predator/cl'
'version_cl.h.tmp' -> 'version_cl.h'
mkdir -p ../cl_build
cd ../cl_build && cmake -D GCC_HOST='/usr/bin/gcc-6.4.0' ../cl
-- The C compiler identification is Clang 6.0.1
-- The CXX compiler identification is Clang 6.0.1
-- Check for working C compiler: /usr/lib/llvm/6/bin/clang-6.0
-- Check for working C compiler: /usr/lib/llvm/6/bin/clang-6.0 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/lib/llvm/6/bin/clang++-6.0
-- Check for working CXX compiler: /usr/lib/llvm/6/bin/clang++-6.0 -- works
[...]
-- GCC_HOST: /usr/bin/gcc-6.4.0
-- GCC_INC_DIR: /usr/lib/gcc/x86_64-pc-linux-gnu/6.4.0/plugin/include
-- 'echo "#include <new>" | /usr/bin/gcc-6.4.0 -m32 -c -xc++ -' works
-- CC1_HOST: /usr/libexec/gcc/x86_64-pc-linux-gnu/6.4.0/cc1
-- checking whether fancy_abort() in /usr/libexec/gcc/x86_64-pc-linux-gnu/6.4.0/cc1 is mangled
-- GCC plug-ins will be built with C++ linkage
[...]
100% tests passed, 0 tests failed out of 191
[...]
100% tests passed, 0 tests failed out of 1152

Total Test time (real) =  14.48 sec
make: Leaving directory '/home/kdudka/git/predator/sl'
CC=clang-6.0 CXX=clang++-6.0 ./switch-host-gcc.sh /usr/bin/gcc-6.4.0: real 58.827, user 3:09.94, sys 16.400, cpu 350%
```

If you are still able to reproduce the crash, please report it as a separate issue with some steps to reproduce.

Comment (versokova) [reply README]: Sorry, I believe you. It looks like another macos issue. The C/CXX compilers were AppleClang 10.0.0.10001044, Clang 6.0.1, Clang 7.0.1 and Clang 8.0.0. Compilation starts with many warnings from gcc plugin headers.

Is in gcc binary anything suspicious?

```
  3/191 Test   #3: compile-self-01-dry-run ..........***Failed    0.54 sec
[...]
/Users/nika/Documents/REPOS/repo-darwin/cl_build/tests/libcl_smoke_test.so: debug: plug-in is compiled against gcc 6.5.0, built at 20181026, conf: ../configure --build=x86_64-apple-darwin17.7.0 --prefix=/usr/local/Cellar/gcc@6/6.5.0 --libdir=/usr/local/Cellar/gcc@6/6.5.0/lib/gcc/6 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-6 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --enable-stage1-checking --enable-checking=release --enable-lto --with-build-config=bootstrap-debug --disable-werror --with-pkgversion='Homebrew GCC 6.5.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --disable-nls --disable-multilib
```


```
[...]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-dev-null' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_typedot.cc:200: debug: ClTypeDotGenerator: created dot file '/dev/null' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl_build/tests/libcl_smoke_test.so: debug: plug-in successfully initialized
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-safe_basename.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-init_plugin_name.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
In function 'void init_plugin_name(const plugin_name_args*)':
cc1plus: warning: unhandled GIMPLE_RESX
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1839: note: raised from 'handle_stmt_resx' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:265:16: warning: unreachable label 'L11' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:279:1: warning: unreachable label 'L14' [-fplugin=libcl_smoke_test.so]
<unknown location>: warning: unreachable label 'L4' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L5' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L8' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:265:16: warning: unreachable label 'L11' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:279:1: warning: unreachable label 'L14' [-fplugin=libcl_smoke_test.so]
<unknown location>: warning: unreachable label 'L4' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L5' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L8' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:265:16: warning: unreachable label 'L11' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:279:1: warning: unreachable label 'L14' [-fplugin=libcl_smoke_test.so]
<unknown location>: warning: unreachable label 'L4' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L5' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L8' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-free_plugin_name.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-dummy_printer.dot' [internal location] [-fplugin=libcl_smoke_test.so]
In function 'void dummy_printer(const char*)':
cc1plus: warning: unhandled GIMPLE_NOP
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1893: note: raised from 'cb_walk_gimple_stmt' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-trivial_printer.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-cl_warn.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-cl_error.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-type_db_hash.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-var_db_hash.dot' [internal location] [-fplugin=libcl_smoke_test.so]
*** WARNING *** there are active plugins, do not report this as a bug unless you can reproduce it without enabling any plugins.
Event                            | Plugins
PLUGIN_FINISH_UNIT               | libcl_smoke_test
PLUGIN_FINISH                    | libcl_smoke_test
PLUGIN_START_UNIT                | libcl_smoke_test
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c: In function 'hashval_t var_db_hash(const void*)':
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:333:18: internal compiler error: Segmentation fault: 11
 static hashval_t var_db_hash(const void *p)
                  ^~~~~~~~~~~
```

Comment (versokova) [reply README]: I got it. On macos there are 2 types of stdlib: libc   used by clang/llvm and libstdc   used by gcc. Which one is set depend on compiler. But if you have dependencies in your code like lib for gcc plugins or llvm passes, you must compile with the corresponding library.

gcc --> gcc [OK]
```
$ otool -L cl_build/tests/libcl_smoke_test.so
cl_build/tests/libcl_smoke_test.so:
	@rpath/libcl_smoke_test.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/gcc@6/lib/gcc/6/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.22.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
	/usr/local/lib/gcc/6/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```

gcc --> clang [BAD]
```
$ otool -L cl_build/tests/libcl_smoke_test.dylib
cl_build/tests/libcl_smoke_test.dylib:
	@rpath/libcl_smoke_test.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/gcc@6/lib/gcc/6/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.22.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
	/usr/local/lib/gcc/6/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```

clang --> clang [OK]
```
$ otool -L cl_build/tests/libcl_smoke_test.dylib
cl_build/tests/libcl_smoke_test.dylib:
	@rpath/libcl_smoke_test.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/llvm@7/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```

clang --> gcc [BAD]
```
$ otool -L cl_build/tests/libcl_smoke_test.so
cl_build/tests/libcl_smoke_test.so:
	@rpath/libcl_smoke_test.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```
I'll upload the commit later.

File: cl/tests/CMakeLists.txt [hunk 16]
```diff
@@ -30,9 +30,9 @@ CL_BUILD_COMPILER_PLUGIN(chk_pt chk_pt_core "")
 add_library(cl_smoke_test_core STATIC cl_smoke_test.cc)
 CL_BUILD_COMPILER_PLUGIN(cl_smoke_test cl_smoke_test_core "")
 
-# get the full paths of plugins
-get_property(VK_PLUG TARGET chk_var_killer PROPERTY LOCATION)
-get_property(PT_PLUG TARGET chk_pt         PROPERTY LOCATION)
+# set the full paths of plugins
+set(VK_PLUG ${cl_BINARY_DIR}/tests/libchk_var_killer${CMAKE_SHARED_LIBRARY_SUFFIX})
+set(PT_PLUG ${cl_BINARY_DIR}/tests/libchk_pt${CMAKE_SHARED_LIBRARY_SUFFIX})
```
Comment (kdudka) [inline cl/tests/CMakeLists.txt]: Why is this needed?

I am afraid that it unnecessarily hard-wires the `lib` prefix, as well as the particular directory layout.

File: sl/CMakeLists.txt [hunk 17]
```diff
@@ -83,8 +83,8 @@ add_library(predator STATIC
 # build compiler plug-in (libsl.so/.dylib)
 CL_BUILD_COMPILER_PLUGIN(sl predator ../cl_build)
 
-# get the full path of libsl.so/.dylib
-get_property(SL_PLUG TARGET sl PROPERTY LOCATION)
+# set the full path of libsl.so/.dylib
+set(SL_PLUG ${cl_BINARY_DIR}/tests/libsl${CMAKE_SHARED_LIBRARY_SUFFIX})
```
Comment (kdudka) [inline sl/CMakeLists.txt]: This does not look correct.  Why would we put the `sl` plug-in into `${cl_BINARY_DIR}/tests`?

**Comment by kdudka** [comment]
I believe it is time to merge this pull request.  Please submit fixes for the remaining issues as a separate pull request.  Thank you for the contribution!
